XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,540 @@
+// completed.
+function destroy(id) { }
+```
+
+
+### `async_wrap`
+
+The object returned from `require('async_wrap')`.
+
+
+#### `async_wrap.currentId()`
+
+Return the id of the current execution context. Useful to track the
+asynchronous execution chain. It can also be used to propagate state without
+needing to use the `AsyncHook` constructor. For example:
+
+```js
+const map = new Map();
+
+net.createServer((c) => {
+  // TODO: In order for this to be useful it needs to be possible to get the
+  // id of a given handle. Otherwise it'll be impossible to store state at
+  // moments like this for later usage using the handle's id.
+}).listen(8111);
+```
+
+
+### Constructor: `AsyncHook(callbacks)`
+
+The `AsyncHook` constructor returns an instance that contains information about
+the callbacks that are to fire during specific asynchronous events in the
+lifetime of the event loop. The focal point of these calls centers around the
+lifetime of `AsyncWrap`. These callbacks will also be called to emulate the
+lifetime of handles and requests that do not fit this model. For example,
+`HTTPParser` instances are recycled to improve performance. So the `destroy()`
+callback would be called manually after a connection is done using it. Just
+before it's placed back into the unused resource pool.
+
+All callbacks are optional. So, for example, if only resource cleanup needs to
+be tracked then only the `destroy()` callback needs to be passed.
+
+**Error Handling**: If any callback throws the application will print the stack
+trace and exit. The exit path does follow that of any uncaught exception,
+except for the fact that it is not catchable by an uncaught exception handler,
+so any `'exit'` callbacks will fire. Unless the application is run with
+`--abort-on-uncaught-exception`. In which case a stack trace will be printed
+and the application will exit, leaving a core file.
+
+The reason for this error handling behavior is that these callbacks are running
+at potentially volatile points in an object's lifetime. For example during
+class construction and destruction. Because of this, it is deemed necessary to
+bring down the process quickly as to prevent an unintentional abort in the
+future. This is subject to change in the future if a comprehensive analysis is
+performed to ensure an exception can follow the normal control flow without
+unintentional side effects.
+
+
+#### `asyncHook.enable()`
+
+Enable the callbacks for a given `AsyncHook` instance. Once a callback fires on
+an asynchronous event they will continue to fire for all nested asynchronous
+events. Even after the instance has been disabled. The following is an example
+of how this operates:
+
+```js
+const async_wrap = require('async_wrap');
+const hooks = new async_wrap.AsyncHook({ init });
+hooks.enable();
+
+net.createServer((c) => {
+  // Even though the hooks have been disabled by the time this callback runs
```

Comment:
Did the behaviour change? Previously one would have to call `.listen()`, as that is what created the handle.